pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentStateResource.java

```diff
@@ -0,0 +1,144 @@
+import org.apache.pinot.controller.helix.core.assignment.segment.SegmentAssignmentUtils;
+import org.apache.pinot.controller.helix.core.realtime.PinotLLCRealtimeSegmentManager;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.config.table.assignment.InstancePartitionsType;
+import org.apache.pinot.spi.utils.CommonConstants;
+import org.apache.pinot.spi.utils.builder.TableNameBuilder;
+
+
+@Api(tags = Constants.SEGMENT_TAG)
+@Path("/")
+public class PinotSegmentStateResource {
+
+  @Inject
+  PinotHelixResourceManager _pinotHelixResourceManager;
+
+  // TODO add support for choosing default offset for other partitions
+  private static class RealtimeSegmentResumeConfig {
+    private Map<Integer, Long> _partitionIdToOffsetMap;
+
+    public Map<Integer, Long> getPartitionIdToOffsetMap() {
+      return _partitionIdToOffsetMap;
+    }
+
+    public void setPartitionIdToOffsetMap(Map<Integer, Long> partitionIdToOffsetMap) {
+      _partitionIdToOffsetMap = partitionIdToOffsetMap;
+    }
+  }
+
+  @POST
+  @Path("/segmentState/{tableName}/resumeRealtimeTable")
+  @Produces(MediaType.APPLICATION_JSON)
+  @Consumes(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Resume a realtime table", notes = "Resume a segment")
+  public String resumeRealtimeTable(
+      @ApiParam(value = "Name of the table", required = true) @PathParam("tableName") String tableName,
+      RealtimeSegmentResumeConfig realtimeSegmentResumeConfig) {
+    TableConfig tableConfig = _pinotHelixResourceManager.getTableConfig(tableName, TableType.REALTIME);
+    String realtimeTableName = TableNameBuilder.REALTIME.tableNameWithType(tableName);
+    InstancePartitions instancePartitions =
+        InstancePartitionsUtils.fetchOrComputeInstancePartitions(_pinotHelixResourceManager.getHelixZkManager(),
+            tableConfig, InstancePartitionsType.CONSUMING);
+    Map<InstancePartitionsType, InstancePartitions> instancePartitionsMap =
+        Collections.singletonMap(InstancePartitionsType.CONSUMING, instancePartitions);
+
+    IdealState idealState =
+        HelixHelper.getTableIdealState(_pinotHelixResourceManager.getHelixZkManager(), realtimeTableName);
+    int numReplicas = PinotLLCRealtimeSegmentManager.getNumReplicas(tableConfig, instancePartitions);
+    Map<String, Map<String, String>> instanceStatesMap = idealState.getRecord().getMapFields();
+
+    for (Map.Entry<Integer, Long> partitionOffset : realtimeSegmentResumeConfig.getPartitionIdToOffsetMap()
+        .entrySet()) {
+      int partitionId = partitionOffset.getKey();
+      Long offset = partitionOffset.getValue();
+
+      // Step 1: Create PROPERTYSTORE nodes for each partitionId (IN_PROGRESS)
+      long newSegmentCreationTimeMs = System.currentTimeMillis();
+
+      // TODO(saurabh) Get the sequence number from ideal state (largest seq # + 1)
+      int seqNumber = Math.abs(new Random().nextInt());
```

Comment:
This to-do should be addressed in the first version

Reply:
Yes thats the plan. Will address

Reply:
Addressed